content/post/2020-08-30-keep.source/index.Rmd

```diff
@@ -0,0 +1,130 @@
+
+The installed packages in the library do not contain the original R files, see [ggplot2 source code](https://github.com/tidyverse/ggplot2/) and ggplot2 on my disk
+
+```{r}
+fs::dir_tree(
+  file.path(
+    find.package("ggplot2"),
+    "R"
+    )
+  )
+```
+
+Under the R folder, there are three files that don't even have the dot R extension!
+
+## How is code stored?
+
+Now, regarding the code, let's mention two important things happening to it.
+
+### Byte compilation
+
+Since [R 3.5](https://cran.r-project.org/doc/manuals/r-release/NEWS.3.html), the code is byte-compiled by default which means it is also stored in a format easier for a machine to deal with.
+You can learn more about byte compilation in the [Efficient R Programming book by Colin Gillespie and Robin Lovelace](https://bookdown.org/csgillespie/efficientR/programming.html#the-byte-compiler), and in [a talk by R Core Member Tomas Kalibera](https://blog.revolutionanalytics.com/2017/08/take-advantage-compiler.html).
+
+### Original formatting and comments?
+
+Also, by default, note that the source code is stripped of all empty lines and comments because they are useless for code execution and take up space.[^thx]
+
+[^thx]: This came to my attention thanks to [a question by Ofek Shilon on RStudio community](https://community.rstudio.com/t/keep-source-pkgs-vs-keep-source-options/69245).
+
+It is similar to CSS, JS, HTML being minified in web development to make websites load faster.
+Now sometimes you might want to keep code with its comments: 
+as an user for being able to read it locally with all its comments,
+as a developer for debugging or profiling (to have line numbers in parsed code refer to actual line numbers you can look up in your scripts).
+
+As an user installing packages, you need to look into the `keep.source.pkgs` option in `options()` that influences the behavior of package installation, or for a specific package you'd write `install.packages("rhub", INSTALL_opts = "--with-keep.source", type = "source")`.[^lookup] If you use Windows or Mac and don't write `type = "source"`, binaries might be use in which case the `keep.source.pkgs` option is ignored.
+
+[^lookup]: When viewing source code you might get a better default experience by [loading `lookup`](https://github.com/jimhester/lookup#default-printing) in your [.Rprofile](https://rstats.wtf/r-startup.html#rprofile) **if you don't use Windows, because if you use Windows and do this you won't be able to update the loaded packages with compiled code**.
+
+As a developer working on a package, you need to make sure the source is kept as is [when building the package, thanks to the `--with-keep.source` option](https://support.rstudio.com/hc/en-us/articles/205612627-Debugging-with-RStudio#debugging-in-packages), and when loading it (lucky you, the relevant `keep.source` option is `TRUE` by default in interactive sessions :tada:).
```

Comment:
Maybe "building" -> "installing", to be closer to the R lingo, but of course binary packages are "built" with `R CMD INSTALL --build`...